File: cpp/best_time_to_buy_and_sell_stock_iii.cc
/*
 * Filename: best_time_to_buy_and_sell_stock_iii.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Wed 05 Aug 2015 06:46:03 PM EDT
 */

#include <climits>
#include <iostream>

using namespace std;

class Solution {
	public:
		int maxProfit(vector<int>& prices)
		{
			if (prices.empty() || prices.size() == 1)
			{
				return 0;
			}

			int n = prices.size();
			vector<int> first (n);
			vector<int> second (n);

			int mini = prices[0];
			int maxi = prices[n-1];
			int res = 0;

			for (int i = 1; i < n; i++) {
				mini = min(mini, prices[i]);
				first[i] = max(first[i-1], prices[i] - mini);
			}

			for (int i = n-2; i >= 0; i--) {
				maxi = max(maxi, prices[i]);
				second[i] = max(second[i+1], maxi - prices[i]);
			}

			for (int i = 0; i < n; i++) {
				res = max(res, first[i] + second[i]);
			}

			return res;
		}
};

File: java/BestTimetoBuyandSellStockIII.java
/*
 * Filename: BestTimetoBuyandSellStockII.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Mon 29 Jun 2015 10:43:59 AM EDT
 */

public class Solution {
	public int maxProfit(int[] prices) {
		if (prices == null || prices.length <= 1) {
			return 0;
		}

		int min = Integer.MAX_VALUE;
		int max = 0;
		int tmpIdx = -1;
		int minIdx = -1;
		int maxIdx = -1;

		for (int i = 0; i < prices.length; i++) {
			int n = prices[i];

			if (n < min) {
				min = n;
				tmpIdx = i;
			}

			if (n - min > max) {
				max = n - min;
				minIdx = tmpIdx;
				maxIdx = i;
			}
		}

		if (maxIdx == -1) {
			return 0;
		}

		int mmin = Integer.MAX_VALUE;
		int mmax = 0;
		for (int i = maxIdx; i >= minIdx; i--) {
			mmin = Math.min(mmin, prices[i]);
			mmax = Math.max(mmax, prices[i] - mmin);
		}

		int mmmin = Integer.MAX_VALUE;
		int mmmax = 0;
		for (int i = 0; i < minIdx; i++) {
			mmmin = Math.min(mmmin, prices[i]);
			mmmax = Math.max(mmmax, prices[i] - mmmin);
		}

		int mmmmin = Integer.MAX_VALUE;
		int mmmmax = 0;
		for (int i = maxIdx + 1; i < prices.length; i++) {
			mmmmin = Math.min(mmmmin, prices[i]);
			mmmmax = Math.max(mmmmax, prices[i] - mmmmin);
		}

		return max + Math.max(mmax, Math.max(mmmax, mmmmax));
	}
}
